cpu/common/parse.c

```diff
@@ -747,19 +741,100 @@ readfile_elf (char *filename)
     }
 
   if (NULL != str_tbl)
+    free (str_tbl);
+
+  if (NULL != sym_tbl)
+    free (sym_tbl);
+}
+
+static void
+readfile_elf (char *filename)
+{
+
+  FILE *inputfs;
+  struct elf32_hdr elfhdr;
+  struct elf32_phdr *elf_phdata = NULL;
+  struct elf32_shdr *elf_shdata;
+  int loaded;
+
+  if (!(inputfs = fopen (filename, "r")))
     {
-      free (str_tbl);
+      perror ("readfile_elf");
+      exit (1);
     }
 
-  if (NULL != sym_tbl)
+  /* Load elf header */
+  if (fread (&elfhdr, sizeof (elfhdr), 1, inputfs) != 1)
     {
-      free (sym_tbl);
+      perror ("readfile_elf");
+      exit (1);
     }
 
-  free (s_str);
-  free (elf_phdata);
-  free (elf_shdata);
+  /* Read in elf section header */
+  if ((elf_shdata =
+       (struct elf32_shdr *) malloc (ELF_SHORT_H (elfhdr.e_shentsize) *
+				     ELF_SHORT_H (elfhdr.e_shnum))) == NULL)
+    {
+      perror ("readfile_elf");
+      exit (1);
+    }
 
+  if (fseek (inputfs, ELF_LONG_H (elfhdr.e_shoff), SEEK_SET) != 0)
+    {
+      perror ("readfile_elf");
+      exit (1);
+    }
+
+  if (fread
+      (elf_shdata,
+       ELF_SHORT_H (elfhdr.e_shentsize) * ELF_SHORT_H (elfhdr.e_shnum), 1,
+       inputfs) != 1)
+    {
+      perror ("readfile_elf");
+      exit (1);
+    }
+
+  /* Read in elf program headers if available */
+  if (ELF_LONG_H (elfhdr.e_phoff))
+    {
+      if ((elf_phdata =
+	   (struct elf32_phdr *) malloc (ELF_SHORT_H (elfhdr.e_phnum) *
+					 ELF_SHORT_H (elfhdr.e_phentsize))) ==
+	  NULL)
+	{
+	  perror ("readfile_elf");
+	  exit (1);
+	}
+
+      if (fseek (inputfs, ELF_LONG_H (elfhdr.e_phoff), SEEK_SET) != 0)
+	{
+	  perror ("readfile_elf");
+	  exit (1);
+	}
+
+      if (fread
+	  (elf_phdata,
+	   ELF_SHORT_H (elfhdr.e_phnum) * ELF_SHORT_H (elfhdr.e_phentsize),
+	   1, inputfs) != 1)
+	{
+	  perror ("readfile_elf");
+	  exit (1);
+	}
+    }
+
+  loaded = load_by_program_headers_elf(inputfs, &elfhdr, elf_phdata);
+  if (loaded <= 0)
+    {
+      loaded = load_by_section_headers_elf(inputfs, &elfhdr,
+					   elf_phdata, elf_shdata);
+    }
+  if (loaded)
+    load_labels_elf(inputfs, &elfhdr, elf_shdata);
```

Comment:
This bit is the core of the new logic.  The `load_by_section_headers_elf` and `load_labels_elf` were split out from this `readfile_elf` function to make it easier to follow.    The `load_by_program_headers_elf` function is new.